File: client/lib/world/island_field.dart
/// Where the floating islands are — twin of `server/resources/js/world/islandField.ts`.
/// The sky is divided into [islandCell]-metre cells; each cell hashes (seed, ix, iz)
/// to decide whether it holds an island and what it looks like. Cell (0, 0) is the
/// legacy island over the spawn pad.
library;

import 'island_gen.dart';
import 'island_template.dart';
import 'js_math.dart';
import 'noise.dart';

const int islandCell = 96;
const double islandChance = 0.5;

/// lowest grass layer of a generated island floats between these heights
const int islandMinY = 80;
const int islandMaxY = 96;
const int legacyIslandY = 80;

/// widest island plus the room its trees need; centres stay this far inside their cell
const int maxHalfExtent = 33;

/// templates kept in memory; islands beyond that are regenerated when needed
const int _templateCache = 48;

const int _saltExists = 1;
const int _saltOffsetX = 2;
const int _saltOffsetZ = 3;
const int _saltSize = 4;
const int _saltHeight = 5;
const int _saltDepth = 6;
const int _saltLake = 7;
const int _saltSeed = 8;
const int _saltBaseY = 9;

final class PlacedIsland {
  const PlacedIsland({
    required this.ix,
    required this.iz,
    required this.x,
    required this.z,
    required this.baseY,
    required this.params,
  });

  final int ix;
  final int iz;

  /// world x/z of the island centre
  final int x;
  final int z;

  /// world y of the island's lowest grass layer (template y = 0)
  final int baseY;
  final IslandParams params;

  String get key => '$ix,$iz';
}

/// The island in a sky cell, or null for an empty cell. Pure in (seed, ix, iz).
PlacedIsland? islandAtCell(int seed, int ix, int iz) {
  if (ix == 0 && iz == 0) {
    return PlacedIsland(
      ix: ix,
      iz: iz,
      x: 0,
      z: 0,
      baseY: legacyIslandY,
      params: islandLarge,
    );
  }
  if (hash01(seed, ix, iz, _saltExists) >= islandChance) return null;
  final size =
      24 + 2 * (hash01(seed, ix, iz, _saltSize) * 17).floor(); // 24..56, even
  const maxJitter = islandCell / 2 - maxHalfExtent;
  final jx = (hash01(seed, ix, iz, _saltOffsetX) * 2 - 1) * maxJitter;
  final jz = (hash01(seed, ix, iz, _saltOffsetZ) * 2 - 1) * maxJitter;
  return PlacedIsland(
    ix: ix,
    iz: iz,
    x: jsRound(ix * islandCell + islandCell / 2 + jx),
    z: jsRound(iz * islandCell + islandCell / 2 + jz),
    baseY:
        islandMinY +
        (hash01(seed, ix, iz, _saltBaseY) * (islandMaxY - islandMinY + 1))
            .floor(),
    params: IslandParams(
      size: size,
      seed: hashInt(seed, ix, iz, _saltSeed),
      maxHeight: 5 + (hash01(seed, ix, iz, _saltHeight) * 5).floor(), // 5..9
      depth: 9 + (hash01(seed, ix, iz, _saltDepth) * 8).floor(), // 9..16
      padRadius: 0,
      lake: hash01(seed, ix, iz, _saltLake) < 0.4,
      treeDensity: 0.022,
    ),
  );
}

/// Islands whose footprint may overlap the block range [x0, x1] × [z0, z1].
/// The player controller passes fractional positions, so the bounds are num.
List<PlacedIsland> islandsNear(int seed, num x0, num z0, num x1, num z1) {
  final out = <PlacedIsland>[];
  // `/` then floor, not `~/`: the range goes negative
  final ix0 = ((x0 - maxHalfExtent) / islandCell).floor();
  final ix1 = ((x1 + maxHalfExtent) / islandCell).floor();
  final iz0 = ((z0 - maxHalfExtent) / islandCell).floor();
  final iz1 = ((z1 + maxHalfExtent) / islandCell).floor();
  for (var ix = ix0; ix <= ix1; ix++) {
    for (var iz = iz0; iz <= iz1; iz++) {
      final island = islandAtCell(seed, ix, iz);
      if (island == null) continue;
      if (island.x + maxHalfExtent < x0 || island.x - maxHalfExtent > x1) {
        continue;
      }
      if (island.z + maxHalfExtent < z0 || island.z - maxHalfExtent > z1) {
        continue;
      }
      out.add(island);
    }
  }
  return out;
}

/// Memoised island templates (generation is ~10 ms for a large island).
final class IslandTemplates {
  final Map<String, IslandTemplate> _cache = {};

  IslandTemplate get(PlacedIsland island) {
    final key = island.key;
    final cached = _cache.remove(key);
    if (cached != null) {
      _cache[key] = cached; // refresh LRU position
      return cached;
    }
    final t = IslandTemplate(island.params);
    if (_cache.length >= _templateCache) _cache.remove(_cache.keys.first);
    _cache[key] = t;
    return t;
  }
}

File: node-server/inertia/world/islandField.ts
/**
 * Where the floating islands are. The sky is divided into ISLAND_CELL-metre cells;
 * each cell hashes (seed, ix, iz) to decide whether it holds an island, where inside
 * the cell it sits, how high it floats and which island params it uses. Islands never
 * leave their cell, so they never overlap. Cell (0, 0) is the legacy island: the
 * original Island_Large, floating over the spawn pad so old saves keep their builds.
 */
import { ISLAND_LARGE, type IslandParams } from './islandGen.ts'
import { IslandTemplate } from './islandTemplate.ts'
import { hash01, hashInt } from './noise.ts'

export const ISLAND_CELL = 96
export const ISLAND_CHANCE = 0.5
/** lowest grass layer of a generated island floats between these heights */
export const ISLAND_MIN_Y = 80
export const ISLAND_MAX_Y = 96
export const LEGACY_ISLAND_Y = 80
/** widest island plus the room its trees need; centres stay this far inside their cell */
const MAX_HALF_EXTENT = 33
/** templates kept in memory; islands beyond that are regenerated when needed */
const TEMPLATE_CACHE = 48

const SALT = { exists: 1, offsetX: 2, offsetZ: 3, size: 4, height: 5, depth: 6, lake: 7, seed: 8, baseY: 9 } as const

export interface PlacedIsland {
  ix: number
  iz: number
  /** world x/z of the island centre */
  x: number
  z: number
  /** world y of the island's lowest grass layer (template y = 0) */
  baseY: number
  params: IslandParams
}

/** The island in a sky cell, or null for an empty cell. Pure in (seed, ix, iz). */
export function islandAtCell(seed: number, ix: number, iz: number): PlacedIsland | null {
  if (ix === 0 && iz === 0) return { ix, iz, x: 0, z: 0, baseY: LEGACY_ISLAND_Y, params: ISLAND_LARGE }
  if (hash01(seed, ix, iz, SALT.exists) >= ISLAND_CHANCE) return null
  const size = 24 + 2 * Math.floor(hash01(seed, ix, iz, SALT.size) * 17) // 24..56, even
  const maxJitter = ISLAND_CELL / 2 - MAX_HALF_EXTENT
  const jx = (hash01(seed, ix, iz, SALT.offsetX) * 2 - 1) * maxJitter
  const jz = (hash01(seed, ix, iz, SALT.offsetZ) * 2 - 1) * maxJitter
  return {
    ix, iz,
    x: Math.round(ix * ISLAND_CELL + ISLAND_CELL / 2 + jx),
    z: Math.round(iz * ISLAND_CELL + ISLAND_CELL / 2 + jz),
    baseY: ISLAND_MIN_Y + Math.floor(hash01(seed, ix, iz, SALT.baseY) * (ISLAND_MAX_Y - ISLAND_MIN_Y + 1)),
    params: {
      size,
      seed: hashInt(seed, ix, iz, SALT.seed),
      maxHeight: 5 + Math.floor(hash01(seed, ix, iz, SALT.height) * 5), // 5..9
      depth: 9 + Math.floor(hash01(seed, ix, iz, SALT.depth) * 8), // 9..16
      padRadius: 0,
      lake: hash01(seed, ix, iz, SALT.lake) < 0.4,
      treeDensity: 0.022,
    },
  }
}

/** Islands whose footprint may overlap the block range [x0, x1] × [z0, z1]. */
export function islandsNear(seed: number, x0: number, z0: number, x1: number, z1: number): PlacedIsland[] {
  const out: PlacedIsland[] = []
  const ix0 = Math.floor((x0 - MAX_HALF_EXTENT) / ISLAND_CELL)
  const ix1 = Math.floor((x1 + MAX_HALF_EXTENT) / ISLAND_CELL)
  const iz0 = Math.floor((z0 - MAX_HALF_EXTENT) / ISLAND_CELL)
  const iz1 = Math.floor((z1 + MAX_HALF_EXTENT) / ISLAND_CELL)
  for (let ix = ix0; ix <= ix1; ix++) {
    for (let iz = iz0; iz <= iz1; iz++) {
      const island = islandAtCell(seed, ix, iz)
      if (!island) continue
      if (island.x + MAX_HALF_EXTENT < x0 || island.x - MAX_HALF_EXTENT > x1) continue
      if (island.z + MAX_HALF_EXTENT < z0 || island.z - MAX_HALF_EXTENT > z1) continue
      out.push(island)
    }
  }
  return out
}

/** Memoised island templates (generation is ~10 ms for a large island). */
export class IslandTemplates {
  private readonly cache = new Map<string, IslandTemplate>()

  get(island: PlacedIsland): IslandTemplate {
    const key = `${island.ix},${island.iz}`
    let t = this.cache.get(key)
    if (t) {
      // refresh LRU position
      this.cache.delete(key)
      this.cache.set(key, t)
      return t
    }
    t = new IslandTemplate(island.params)
    if (this.cache.size >= TEMPLATE_CACHE) this.cache.delete(this.cache.keys().next().value!)
    this.cache.set(key, t)
    return t
  }
}
